autorel/changelog_generator/changelog_gen.py

```diff
@@ -0,0 +1,279 @@
+## Main changelog generation logic goes here
+import pygit2
+from parser import Parser
+from parser_config import PARSING_FUNCTIONS
+from structures import (PullRequest,
+                        Issue,
+                        Contributor
+                        )
+from settings import (PROJECT,
+                      CLONE_URL
+                      )
+from github_plugin import (getPull,
+                           getCommitList,
+                           getIssue
+                           )
+
+
+class ChangelogEntry(object):
+    """
+        Defines a changelog entry
+    """
+    def __init__(self, category, contributor, url, text):
+        self._category = category
+        self._contributor = contributor
+        self._url = url
+        self._text = text
+
+    @property
+    def url(self):
+        return self._url
+
+    @property
+    def text(self):
+        return self._text
+
+    @property
+    def contributor(self):
+        return self._contributor
+
+    @property
+    def category(self):
+        return self._category
+
+
+class ChangelogGenerator(object):
+    """
+        The Changelog Generator class
+    """
+    # The generated changelog will contain the following categories
+    categories = (
+        "enhancement",
+        "bugfix",
+        "fixed_issue",
+        "merged_pull"
+    )
+
+    def __init__(self):
+        self._user,self._repoName = tuple(PROJECT.split('/'))
+        # A dict of ChangelogEntry objects
+        self._entries = {}
+        # Initialise the supported categories
+        for entry in self.categories:
+            self._entries[entry] = []
+        # List of commit objects for the repo
+        self._commits = []
+        # A dict to keep track of traversal of commits
+        self._commit_map = {}
+
+    def _configure_parser(self):
+        """
+            Configure the parser for different paring funtions
+        """
+        self._parser = Parser(["issues","merges"])
+        for type_ in PARSING_FUNCTIONS.keys():
+            for parsing_func in PARSING_FUNCTIONS[type_]:
+                self._parser.addParser(type_,parsing_func)
+
+    def _clone_repository(self):
+        """
+            Clone and instanstiate a repo instance
+        """
+        clone_path = '/tmp/{0}'.format(self._repoName)
```

Comment:
Actually, I was testing things. So, to avoid re-cloning I computed the path of repo. This will be fixed. The clone will be done in a `SyslogNgRelease` class method and the path will be supplied to the changelog generator.